cpp/src/kmeans/sg_impl.cuh

```diff
@@ -500,39 +516,80 @@ void fit(const ML::cumlHandle_impl &handle, const KMeansParams &params,
   // Device-accessible allocation of expandable storage used as temorary buffers
   MLCommon::device_buffer<char> workspace(handle.getDeviceAllocator(), stream);
 
-  if (params.init == KMeansParams::InitMethod::Random) {
-    // initializing with random samples from input dataset
-    LOG(handle, params.verbose,
-        "KMeans.fit: initialize cluster centers by randomly choosing from the "
-        "input data.\n");
-    initRandom(handle, params, data, centroidsRawData);
-  } else if (params.init == KMeansParams::InitMethod::KMeansPlusPlus) {
-    // default method to initialize is kmeans++
-    LOG(handle, params.verbose,
-        "KMeans.fit: initialize cluster centers using k-means++ algorithm.\n");
-    initKMeansPlusPlus(handle, params, data, centroidsRawData, workspace);
-  } else if (params.init == KMeansParams::InitMethod::Array) {
-    LOG(handle, params.verbose,
-        "KMeans.fit: initialize cluster centers from the ndarray array input "
-        "passed to init arguement.\n");
+  auto n_init = km_params.n_init;
+  if (km_params.init == KMeansParams::InitMethod::Array && n_init != 1) {
+    LOG(handle, km_params.verbose,
+        "Explicit initial center position passed: performing only one init in "
+        "k-means instead of n_init=%d",
+        n_init);
+    n_init = 1;
+  }
 
-    ASSERT(centroids != nullptr,
-           "centroids array is null (require a valid array of centroids for "
-           "the requested initialization method)");
+  std::mt19937 gen(km_params.seed);
+  inertia = std::numeric_limits<DataT>::max();
 
-    centroidsRawData.resize(params.n_clusters * n_features, stream);
-    MLCommon::copy(centroidsRawData.begin(), centroids,
-                   params.n_clusters * n_features, stream);
+  // run k-means algorithm with different seeds
+  for (auto seed_iter = 0; seed_iter < n_init; ++seed_iter) {
+    // generate KMeansParams with different seed
+    KMeansParams params = km_params;
+    params.seed = gen();
```

Comment:
Can we make this deterministic so that the same initial seed will provide the same results? Would somehting like `params.seed = params.random_state + seed_iter`?

Reply:
The generator is initialized as `std::mt19937 gen(km_params.seed);` with the user provided random state (see line 528). This should make it deterministic, are you seeing it otherwise?

Reply:
@akkamesh, I was seeing slightly different results, which is what prompted me to look through the code for the possible cause. I just committed a pytest that runs two kmeans training sessions with the same `random_state` and `n_init=10`. 

The test fails and the resulting cluster centers are quite different. If I set `n_init=1` the pytest passes, but it doesn't pass with `n_init=10`. Once this pytest is passing, I think this PR is ready to go!